caveclient/skeletonservice.py

```diff
@@ -0,0 +1,164 @@
+
+from .auth import AuthClient
+from .base import (
+    ClientBase,
+    _api_endpoints,
+    handle_response,
+)
+from .endpoints import (
+    default_global_server_address,
+    skeletonservice_api_versions,
+    skeletonservice_common,
+)
+from .endpoints import skeletonservice_api_versions, skeletonservice_common
+
+SERVER_KEY = "skeleton_server_address"
+
+
+'''
+Usage
+'''
+
+def SkeletonClient(
```

Comment:
I'm a proponent of avoiding the factory syntax in future code now that each subclient can reason about which version of the service it is talking to, and code paths can be protected with the decorator. I believe this is something we briefly discussed when I implemented that.

Main reasons being that I find this makes autocomplete much worse (since the editor can't reason about exactly what type it will get back when there is MaterializationClientV2 and MaterializationClientV3 etc.), makes documentation more annoying (I hate what I had to do  for the current docs), and makes code harder to read.

But I wonder what @fcollman @ceesem and others think?

Reply:
Agreed — I like the version-limited approach much much more.

Reply:
I thought I had followed a pattern from another example in the repo, as a template off which to create this new service. For example, it looks like ChunkedGraphClient or InfoServiceClient. What is the recommended alternative? I thought I just replicated the code of the existing services.

Reply:
Yeah, that's why Ben was saying it for future code like this we should follow the new pattern (and really we should sit down and do a refactor to incorporate it everywhere).

An example of the pattern can be found here:
https://github.com/CAVEconnectome/CAVEclient/blob/9f365b23b77d76a18b14f80659c527d1c90a47c9/caveclient/chunkedgraph.py#L812

Basically, you decorate your client class functions with a version check (this can be done at the method or the keyword level). In your case, there's no need to have any version checks yet so you can just write a bare SkeletonClient object I think.

Reply:
Okay, I'll look into that.

Reply:
I see the decoration you pointed out, but that class still has `") -> "ChunkedGraphClientV1":` in the `ChunkedGraphClient` definition and a `client_mapping` at the bottom. Is there an example of CaveClient service that follows the intended pattern?

I tried simply removing the factory function entirely, but it was populating a few default parameters for the generated class's ctor, namey: 'auth_header', 'api_version', 'endpoints', and 'server_name'. How do I specify those in the SkeletonClient ctor?

Reply:
I believe this has been properly redesigned in the latest commit. Please confirm.